### PR TITLE
Resource manager2

### PR DESCRIPTION
**Pull request by DjWarmonger**
Branch is now functional and ready for merge.

### CODE REVIEW

**Comment by alexvins** [comment]
No it`s not ready, I contain even more mess in history

**Comment by DjWarmonger** [comment]
Where exactly? Shows 4 files changed here^

**Comment by alexvins** [comment]
No after I fixed history it shows what it should

**Comment by alexvins** [comment]
Now it is rebased properly. Next part is fix build - there are link errors

**Comment by DjWarmonger** [comment]
Well, Travis compiles gcc well. No idea about other platforms.

**Comment by alexvins** [comment]
New files are missing in CMakeLists

**Comment by DjWarmonger** [comment]
I opened this CMakeLists and have no idea what's going on, sorry ¯\_(ツ)_/¯

**Comment by alexvins** [comment]
there is a section in AI\VCAI\CMakeLists.txt
```
set(VCAI_SRCS
		StdInc.cpp

		AIUtility.cpp
		Fuzzy.cpp
		Goals.cpp
		main.cpp
		VCAI.cpp
)
```

just add missing sources

**Comment by alexvins** [comment]
```
../AI/VCAI/ResourceManager.cpp:83:11: error: non-const lvalue reference to type 'ResourceObjective' cannot bind to a temporary of type 'ResourceObjective'
                tryPush(ResourceObjective(res, goal));
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by DjWarmonger** [comment]
Good catch!

File: AI/VCAI/AIUtility.cpp [hunk 1]
```diff
@@ -475,6 +480,50 @@ void getVisibleNeighbours(const std::vector<int3> & tiles, std::vector<int3> & o
...
+	}
+	return ci;
+}
+
+ui64 howManyReinforcementsCanBuy(HeroPtr h, const CGTownInstance * t)
+{
+	ui64 aivalue = 0;
+
+	int freeHeroSlots = GameConstants::ARMY_SIZE - h->stacksCount();
+	for (auto const dc : t->creatures)
```
Comment (alexvins) [inline AI/VCAI/AIUtility.cpp]: no space after `for` and `if` please

File: AI/VCAI/AIhelper.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include "StdInc.h"
+
+#include "AIhelper.h"
+#include "ResourceManager.h"
+
+boost::thread_specific_ptr<AIhelper> ah;
+
+AIhelper::AIhelper()
+{
+	resourceManager.reset(new ResourceManager);
```
Comment (alexvins) [inline AI/VCAI/AIhelper.cpp]: should be `new ResourceManager()`

File: AI/VCAI/AIhelper.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	//TODO: consider common interface with Resource Manager?
+	bool canAfford(const TResources & cost) const;
+	TResources reservedResources() const override;
+	TResources freeResources() const override;
+	TResource freeGold() const override;
+	TResources allResources() const override;
+	TResource allGold() const override;
+
+	Goals::TSubgoal whatToDo(TResources &res, Goals::TSubgoal goal) override;
```
Comment (alexvins) [inline AI/VCAI/AIhelper.h]: put space around `&`

File: AI/VCAI/Goals.cpp [hunk 4]
```diff
@@ -798,25 +874,158 @@ TSubgoal DigAtTile::whatToDoToAchieve()
...
+			if (t.valid()) //we know any path at all
+			{
+				if (ai->isTileNotReserved(h, t)) //no other hero wants to conquer that tile
+				{
+					if (isSafeToVisit(h, dest))
+					{
+						if (dest != t) //there is something blocking our way
+							ret.push_back(sptr(Goals::ClearWayTo(dest, h).setisAbstract(true)));
+						else
+						{
```
Comment (alexvins) [inline AI/VCAI/Goals.cpp]: remove {} here

**Comment by DjWarmonger** [comment]
vcmitest project needs now to link VCAI.

**Comment by alexvins** [comment]
../test/vcai/ResurceManagerTest.cpp:36:8: error: no template named 'make_unique' in namespace 'std'; did you mean simply 'make_unique'?

**Comment by DjWarmonger** [comment]
I meant this one: https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique

**Comment by alexvins** [comment]
It is c++14, we have `make_unique` in global namespace (see Global.h)

**Comment by DjWarmonger** [comment]
ok, will fix

**Pull request merged**